Routing

LeiaQuesada opened the pull request:
## Refactor code 
Moved components into their own files, and into the components directory. This is to address separation of concerns. Also moved React api calls out of components and into apiClient file.

[This branch has been deployed from heroku](https://getmygarden.herokuapp.com/)

Not on the trello board, this is just to address bugs I introduced while trying to refactor after successfully rendering content from database and external api call.

At this point, 
- the user can still log in
- will be prompted to update their profile
- can not get plant recommendations (will fix this shortly)
 
I have scrapped the indoor option for now.

LeiaQuesada committed:
install react-router-dom for rendering page views

LeiaQuesada committed:
Create initial routes

Landing page now has welcome screen with some app info, after authentication, user will see navbar with login info in header. User can change their profile information.

LeiaQuesada committed:
Update npm

Remove react-router-dom from root directory, and upgrade npm.

LeiaQuesada committed:
Redirect path to profile

Improve user experience, redirect to profile form after authenticating.

LeiaQuesada committed:
Declare routes

Specify routes and Components to display. Remove list items and unused links from header/navbar.

LeiaQuesada committed:
Create recommendations route.

Access route via useHistory and useLocation hook to pass props.

LeiaQuesada committed:
Refactor components.

Move React components based on separation of concerns. Navbar components are now in header, content is displaying under header. Authentication functioning properly throughout routes.

LeiaQuesada committed:
To protect routes with Auth0

use auth0 hooks, follow https://github.com/auth0/auth0-react/blob/master/EXAMPLES.md#1-protecting-a-route-in-a-react-router-dom-app

LeiaQuesada committed:
Remove authentication conditionals

Components now checked at App.js, no longer need conditionals whether or not the user is authenticated.

LeiaQuesada committed:
Remove profile path from redirect

LeiaQuesada committed:
Rename files for clarity.

Address file names: SignIn to Home. Remove unused imports.

LeiaQuesada committed:
Refactor code for dynamic content.

Separate out api requests into apiClient. Add conditional to make the home screen display option to update the user profile only if the user has never updated their information. After login, apply condition to show plot data.

LeiaQuesada committed:
Add field to the user’s table.

Based on record value for showing data from previous commit.

LeiaQuesada committed:
Add plant recommendations.

Create link from the home page to allow the user to route to their plant recommendations.

LeiaQuesada committed:
Create junction table.

To allow user to save plants to assigned containers. Key constraints added for table referential integrity.

LeiaQuesada committed:
Begin add plant feature.

Add SQL command to add the user's plant selection to the database. Add container name field to database. Add React router functionality to express server routes for json parsing. Add button to each plant displayed for user to select.

carbonsoda commented on app/src/App.js:
```diff
@@ -1,52 +1,42 @@
-      <UserProfile />
-      <footer>testing</footer>
-    </>
-  ) : (
```
Woah, nice cleanup with this!

carbonsoda commented on app/src/components/Header.js:
```diff
@@ -0,0 +1,34 @@
+  const { logout, user, isAuthenticated } = useAuth0();
+
+  return (
+    <div className="header">
```
It might be a good idea to use a more semantic tag ie `<header>` vs `<div className="header">`. Or you seem to use this component like a navbar, so  `<nav>`  would work too! Up to you on which to use.

LeiaQuesada replied:
Thanks, @carbonsoda!

LeiaQuesada replied:
is it strange to say `<nav className="header">` for styling?

aedward8 replied:
@LeiaQuesada I think thats fine!
https://react-bootstrap.github.io/components/navbar/

Looking at this bootstrap documentation they do that exact thing 
` <Nav className="mr-auto">`

carbonsoda commented on app/src/components/Header.js:
```diff
@@ -0,0 +1,34 @@
+        <>
+          <main>
+            <nav></nav>
+          </main>
```
`<main>` indicates the main body of the page like your Home or User Profile pages. Lines 15-17 should be taken out, since they don't have any effect here

carbonsoda commented on server/db.mjs:
```diff
@@ -33,6 +33,13 @@ export const getPlants = async (zone) =>
+  await db.any("INSERT INTO containers (user_id, plant_id) VALUES ($1, $2);", [
+    container.userid,
+    container.plantid,
+  ])[0];
```
What does the `[0]` mean here? Does it return the entire new row back?

LeiaQuesada replied:
I believe it's the index position since db.any returns an array, i just want the first record returned

carbonsoda commented on app/src/components/Home.js:
```diff
@@ -0,0 +1,63 @@
+            and provides gardening recommendations based on your selections.
+            This way, you can focus on the more relaxed motions of building and
+            watching your efforts literally come to fruition.
+          </p>
```
This is a really nice intro :) Written in friendly language and nice to see on the landing/new user page

carbonsoda commented on app/src/index.js:
```diff
@@ -2,26 +2,27 @@ import React from "react";
 import "./styles.css";
 
-import App from "./App";
+import App, { history } from "./App";
```
Totally optional and just for your info -- 
This would also work `import { App, history } from './App';` if you remove the `default` label from `App.js`
https://github.com/LeiaQuesada/MyGarden/blob/0484e3052478434e2f516591f223f40a07ed3b25/app/src/App.js#L18

It's like the same thing going on in `db.mjs`, where you _could_ import just a few, or import them all like how we do with `import * as db from 'db.mjs'`

carbonsoda commented on app/src/styles.css:
```diff
@@ -70,6 +65,22 @@ button {
   font-size: 1rem;
   width: 100%;
+  color: #30485f;
+  font-family: -apple-system, BlinkMacSystemFont, "Roboto", "Oxygen", "Ubuntu",
```
Since this is for `<img id="userpic">`, the font attributes don't seem to do anything. If a user doesn't have a profile picture, it will generate one for them too.

carbonsoda replied:
With regards to the comment 2 lines up, you could shorten this CSS block to:
```css
#userpic {
  width: 5rem; 
  margin: 0 0 0.2rem 0;
  display: inline-block; 
}
```
`display:inline-block` might also be okay to drop, but I'll leave that to you!

carbonsoda commented on app/src/styles.css:
```diff
@@ -70,6 +65,22 @@ button {
   display: inline-block;
   font-size: 1rem;
   width: 100%;
```
When deployed, this doesn't seem to do anything and just shows `width: 75px` in dev tools.

With that being said, one more note, you use `rem` except for the width. Converting `width` to `rem` might be a good idea! On my screen `5rem` seems to look the exact same as `75px`.

carbonsoda reviewed (approved):
You did a really nice job with refactoring, both with splitting up pieces and renaming. It's very clear from your naming what is the purpose, which is the ideal and reduces # of comments needed since they're self-explanatory 💯 

There are just some small nits with HTML/CSS in the header mainly but that's about it! Those can be addressed later too.

LeiaQuesada committed:
Install material-ui library

Implement suggested changes for semantic tags and styling rendering. Improve display of plant recommendations.

LeiaQuesada committed:
Update README

Reflect project completion features, remove outdated information.

alexcheva commented on README.md:
```diff
@@ -103,42 +103,36 @@ npm run psql
 
-- Login with Facebook or Google account
-- Display facts about user’s zone, based on user’s zipcode input
+- Easy secure login with Google account or Sign Up.
```
👍

aedward8 commented on app/src/App.js:
```diff
@@ -1,52 +1,42 @@
-  } = useAuth0();
+const ProtectedRoute = ({ component, ...args }) => (
+  <Route component={withAuthenticationRequired(component)} {...args} />
+);
```
Interesting is this your react router from react-router-dom? Does this make all the components sent into it a route ?

LeiaQuesada replied:
this is an Auth0 protected route with react-router-dom, if you're using both i can send you the documentation. This was really difficult for me to get going. For Auth0 routes, you have to maintain authentication. We have to pass the args in an Auth0 method.

aedward8 commented on app/src/App.js:
```diff
@@ -1,52 +1,42 @@
+          />
+        </Switch>
+      </Router>
+      <footer>Disclaimer © 2021</footer>
```
This is just a nice to have! But you can also make your footer a component if you wanted to add more things to it

alexcheva commented on app/src/App.js:
```diff
@@ -1,52 +1,42 @@
-      </div>
-      <footer>testing</footer>
+      <Header />
+      {error && <div>{error.message}</div>}
```
Nice! I should try that ❤️

alexcheva commented on app/src/apiClient.js:
```diff
@@ -16,24 +14,32 @@ export const addTask = async (name) => {
-//    });
-//    return await response.json();
-// };
+export const getUser = async (token, email) => {
+  const response = await fetch(`/api/user/${email}`, {
+    headers: {
+      Authorization: `Bearer ${token}`,
+    },
+  });
+  return await response.json();
+};
```
Wow. That's cool.

alexcheva commented on app/package.json:
```diff
@@ -29,18 +29,20 @@
-    "@testing-library/react": "^11.1.0",
-    "@testing-library/user-event": "^13.0.2",
-    "eslint-config-prettier": "^8.2.0",
+    "@testing-library/jest-dom": "^5.12.0",
```
Nice!

aedward8 commented on app/src/apiClient.js:
```diff
@@ -16,24 +14,32 @@ export const addTask = async (name) => {
 
 export const getPlants = async (token, zone) => {
-  const response = await fetch(`/api/plants/${zone}`, {
+  const response = await fetch(`/api/plant/${zone}`, {
```
just a small observation. You changed the api route from plants to plant I think to be consistent with the add plant route below. However this api/plant is in the getPlants function so maybe need to change function to getPlant singular to keep consistent?

aedward8 commented on app/src/apiClient.js:
```diff
@@ -16,24 +14,32 @@ export const addTask = async (name) => {
+    body: JSON.stringify({ userid, plantid }),
+  });
+  return await response.json();
+};
```
Nice your routes are very clear for a reader like me to understand

alexcheva commented on server/server.mjs:
```diff
@@ -45,11 +45,21 @@ user.post("/", async (request, response) => {
+const plant = express.Router();
+plant.use(express.json());
+
+plant.get("/:zone", async (request, response) => {
+  const plants = await db.getPlants(request.params.zone);
```
sweet!

alexcheva commented on app/src/components/UserProfile.js:
```diff
@@ -83,7 +84,7 @@ export default function UserProfile() {
                   username && username !== ""
                     ? username
-                    : email.slice(0, email.indexOf("@"))
+                    : user.email.slice(0, user.email.indexOf("@"))
```
Smart!

alexcheva commented on app/src/components/PlantRecommendations.js:
```diff
@@ -1,36 +1,60 @@
 
 import "../styles.css";
 import { useAuth0 } from "@auth0/auth0-react";
+import { Button, Card } from "@material-ui/core";
```
Nice one!

alexcheva reviewed (approved):
Really great job, Leia!

aedward8 commented on app/src/components/PlantRecommendations.js:
```diff
@@ -1,36 +1,56 @@
           );
         })}
       </ul>
-      <button onClick={getPlants}>Get Recommended Zone Plants</button>
```
Is the getPlants recommendation button moved to another component now?

aedward8 commented on app/src/components/PlantRecommendations.js:
```diff
@@ -1,36 +1,56 @@
+                id="plantcardbutton"
+                onClick={() => handleAddPlant(plant.id)}
+              >
+                Add
```
Is this add button saving plant recommendations that you like?

LeiaQuesada replied:
Yes

aedward8 commented on app/src/components/UserProfile.js:
```diff
@@ -18,30 +15,32 @@ export default function UserProfile() {
   const updateUser = async (event) => {
     event.preventDefault();
-    // check input value for zipcode with regex pattern
     if (/^\d{5}$/.test(zipcode) === false) {
```
awesome way to validate an input!

aedward8 reviewed (approved):
Good Job Leia, your code is looking clear & refactored and is easy to understand while reading through. Currently the login page /landing page works! Im sure your next working on the recommendations page

alexcheva commented on app/src/components/Home.js:
```diff
@@ -0,0 +1,63 @@
+
+  useEffect(() => {
+    async function showProfile() {
+      const token = await getAccessTokenSilently();
```
Is this all that required for the token? 🤔

LeiaQuesada replied:
Yea the useAuth hook and making sure to saving the return value inside an async with the await keyword .. This was the end product that stopped bugging out.

alexcheva reviewed (commented):
looking good
